### PR TITLE
Bozin Dmitry Lab2 Var2

### PR DESCRIPTION
**Pull request by Nilemos**

### CODE REVIEW

File: llvm/labs/lab2/bozin_dmitry/bozin_pluginlab2.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+          if (ValueMap.count(User->getOperand(i))) {
+            User->getOperand(i)->replaceAllUsesWith(
+                ValueMap[User->getOperand(i)]);
+          }
+        }
+      }
+
+      // Handling branch instructions
+      llvm::BasicBlock *PostCallBBNext = PostCallBB->getNextNode();
+      if (PostCallBBNext) {
```
Comment (m-ly4) [inline llvm/labs/lab2/bozin_dmitry/bozin_pluginlab2.cpp:103]:
```suggestion
      if (llvm::BasicBlock *PostCallBBNext = PostCallBB->getNextNode()) {
```

File: llvm/test/lab2/bozin_dmitry/test.ll [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+; CHECK-NEXT:   %cmp = icmp slt i32 %0, 1
+; CHECK-NEXT:   br i1 %cmp, label %if.then, label %if.end
+; CHECK: if.then:                                          ; preds = %entry
+; CHECK-NEXT:   br label %post-call
+; CHECK: 1:                                                
+; CHECK-NEXT:   %2 = alloca float, align 4
+; CHECK-NEXT:   store float 1.000000e+00, ptr %2, align 4
+; CHECK-NEXT:   %3 = load float, ptr %2, align 4
+; CHECK-NEXT:   %4 = fcmp olt float %3, 2.000000e+00
+; CHECK-NEXT:   br label %5
```
Comment (m-ly4) [inline llvm/test/lab2/bozin_dmitry/test.ll:210]: Looks like conditional break is lost. Fix it please.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab2/bozin_dmitry/test.ll [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  store i32 %inc, ptr %a, align 4
+  ret void
+}
+
+; CHECK: define dso_local void @_Z4bar1v() {
+; CHECK-NEXT: entry:
+; CHECK-NEXT:   %a = alloca i32, align 4
+; CHECK-NEXT:   store i32 0, ptr %a, align 4
+; CHECK-NEXT:   br label %post-call
+; CHECK: 0:
```
Comment (aobolensk) [inline llvm/test/lab2/bozin_dmitry/test.ll:40]:
```suggestion
; CHECK: 0:
```
here and in other places

File: llvm/labs/lab2/bozin_dmitry/bozin_pluginlab2.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        CurrentBB =
+            llvm::BasicBlock::Create(F.getContext(), "", &F, PostCallBB);
+        ValueMap[&CalleeBB] = CurrentBB;
+
+        Builder.SetInsertPoint(CurrentBB);
+        for (llvm::Instruction &Inst : CalleeBB) {
+          if (!Inst.isTerminator()) {
+            llvm::Instruction *NewInst = Inst.clone();
+            if (!NewInst) {
+              llvm::errs() << "Error: Failed to clone instruction.\\n";
```
Comment (aobolensk) [inline llvm/labs/lab2/bozin_dmitry/bozin_pluginlab2.cpp:51]:
```suggestion
              llvm::errs() << "Error: Failed to clone instruction.\n";
```

File: llvm/labs/lab2/bozin_dmitry/bozin_pluginlab2.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        PrevBB = CurrentBB;
+      }
+
+      if (PrevBB) {
+        if (PrevBB->getTerminator()) {
+          PrevBB->getTerminator()->eraseFromParent();
+        }
+        Builder.SetInsertPoint(PrevBB);
+        Builder.CreateBr(PostCallBB);
+      }
```
Comment (aobolensk) [inline llvm/labs/lab2/bozin_dmitry/bozin_pluginlab2.cpp:76]: Get rid of this segment duplication

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**